consul/src/main/java/com/linecorp/armeria/client/consul/ConsulEndpointGroup.java

```diff
@@ -0,0 +1,128 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client.consul;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.client.ClientFactory;
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.client.endpoint.DynamicEndpointGroup;
+import com.linecorp.armeria.client.endpoint.EndpointGroup;
+import com.linecorp.armeria.client.retry.Backoff;
+import com.linecorp.armeria.common.consul.ConsulClient;
+
+/**
+ * A Consul-based {@link EndpointGroup} implementation. This {@link EndpointGroup} retrieves the list of
+ * {@link Endpoint}s from a Consul using {@link ConsulClient} and updates it when the status of the
+ * service by polling.
+ */
+public final class ConsulEndpointGroup extends DynamicEndpointGroup {
+
+    /**
+     * Create a {@code ConsulEndpointGroup} with only service name. This {@code ConsulEndpointGroup} will
+     * retrieves the list of {@link Endpoint}s from a local Consul agent(using default consul service port).
+     */
+    public static ConsulEndpointGroup of(String serviceName) {
+        return builder(serviceName).build();
+    }
+
+    /**
+     * Creates a {@code ConsulEndpointGroupBuilder} to build {@code ConsulEndpointGroup}.
+     */
+    public static ConsulEndpointGroupBuilder builder(String serviceName) {
+        return new ConsulEndpointGroupBuilder(serviceName);
+    }
+
+    private final ConsulClient consulClient;
+    private final String serviceName;
+    private final ClientFactory clientFactory;
+    private final Backoff retryBackoff;
+
+    @Nullable
+    private volatile ScheduledFuture<?> scheduledFuture;
+
+    /**
+     * Create a Consul-based {@link EndpointGroup}, endpoints will be retrieved by service name using
+     * {@link ConsulClient}.
+     *
+     * @param consulClient   a consul client
+     * @param serviceName a service name to retrieve
+     * @param clientFactory a client factory
+     * @param retryBackoff a retry backoff to check
+     */
+    ConsulEndpointGroup(ConsulClient consulClient, String serviceName, ClientFactory clientFactory,
+                        Backoff retryBackoff) {
+        this.consulClient = consulClient;
+        this.serviceName = serviceName;
+        this.clientFactory = clientFactory;
+        this.retryBackoff = retryBackoff;
+
+        clientFactory.whenClosed().thenRun(this::closeAsync);
+        initialCheck();
+        whenReady().join();
+        startCheck();
+    }
+
+    private void initialCheck() {
+        consulClient.healthyEndpoints(serviceName)
+                    .thenApply(endpoints -> {
+                        setEndpoints(endpoints);
+                        return endpoints;
+                    })
+                    .thenAccept(whenReady()::complete);
```

Comment:
`setEndpoints` completes `initialEndpointsFuture`.
https://github.com/line/armeria/blob/74cb4a51214515d125f33afac6d76c74edf6c7df/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java#L168

I don't think we need to call `whenReady()::complete`.

How about:
```java
consulClient.healthyEndpoints(serviceName)
           .thenAccept(this::setEndpoints);
```
And we could inline this into constructor.